File: solutions/3363-most-frequent-ids/most-frequent-ids.java
class Solution {
    public long[] mostFrequentIDs(int[] nums, int[] freq) {
        int n = nums.length; 
        Map<Long, Long> cnt = new HashMap(); 
        Queue<long[]> pq = new PriorityQueue<>((x, y) -> Long.compare(y[0], x[0])); 
        long[] ans = new long[n]; 
        for (int i = 0; i < n; ++i) {
            long x = nums[i], f = freq[i]; 
            cnt.merge(x, f, Long::sum); 
            while (!pq.isEmpty() && cnt.get(pq.peek()[1]) != pq.peek()[0]) pq.poll(); 
            pq.add(new long[]{cnt.get(x), x}); 
            ans[i] = pq.peek()[0]; 
        }
        return ans; 
    }
}

File: solutions/3363-most-frequent-ids/most-frequent-ids.cpp
class Solution {
public:
    vector<long long> mostFrequentIDs(vector<int>& nums, vector<int>& freq) {
        priority_queue<pair<long long, int>> pq; 
        unordered_map<int, long long> cnt; 
        vector<long long> ans; 
        for (int i = 0, n = nums.size(); i < n; ++i) {
            int x = nums[i], f = freq[i]; 
            cnt[x] += f; 
            while (pq.size() && cnt[pq.top().second] != pq.top().first) pq.pop(); 
            pq.emplace(cnt[x], x); 
            ans.push_back(pq.top().first); 
        }
        return ans; 
    }
};